Broker updates for Adal Android

jennyf19 opened the pull request:
While testing w/broker, I noticed that if the redirectUri is not correct, we don't throw an exception. The intent is returned as null and no error message is returned. @SomkaPe and I discussed this and decided it was best to throw early since the authentication cannot continue w/an incorrect redirect uri. I added an Android specific error message, and in the logs, we share what the correct uri should be for that application. This is a better user experience because now, the application receives an `AdalException` with a clear error message, compared with nothing before.

Platform specific code here, so not sure how to add unit test coverage, but I did test w/Android device OnePlus.

jennyf19 committed:
remove embedded webview as default in test app, we already have this in the web browser section. will add buttons to the app once we have finalized the implementation

jennyf19 committed:
fix spelling

jennyf19 committed:
additions to test app
throwing for incorrect redirect uri for broker flow

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/embeddedTestApp

jennyf19 commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -37,6 +37,13 @@ public class SecondPage : ContentPage
 
+        static string ClientId = "de49ddaf-c7f8-4a06-8463-3c6ae124fe52";
+
+        //static string AndroidBrokerRedirectURI = "msauth://com.microsoft.xformsdroid.adal/mJaAVvdXtcXy369xPWv2C7mV674=";
```
this is the redirectUri for broker for this application.

henrik-me replied:
any reason this is not either left there or simply deleted? Ideally it should be kept and a button in the ui should be able to switch between broker and non-broker. thoughts?

---
In reply to: [203584400](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r203584400) [](ancestors = 203584400)

jennyf19 replied:
That's a good idea to include a button. I will do that.

jennyf19 replied:
done.

jennyf19 commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/BrokerHelper.cs:
```diff
@@ -210,6 +210,11 @@ public void AcquireToken(IDictionary<string, string> brokerPayload)
                         RequestContext.Logger.ErrorPii(e);
                     }
                 }
+                else
```
we were not doing anything with a null intent here. Maybe this exception is too specific? Could just be "null intent" as an error message, but one needs to check ADAL logs to know the redirect uri was incorrect. thoughts?

henrik-me replied:
Precise exceptions are good for supportability!

Aside: Seems to follow our existing ADAL pattern. One thing though to consider changing for MSAL exceptions. Looking forward to the discussions on exceptions for MSAL later.

---
In reply to: [203584606](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r203584606) [](ancestors = 203584606)

jennyf19 replied:
agree.

SomkaPe replied:
Yes this exception is to specific here, it might be any exception - not necessary related to incorrect URL,

---
In reply to: [203584606](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r203584606) [](ancestors = 203584606)

SomkaPe replied:
Exception should be thrown in the place were it is generated

henrik-me replied:
does broker throw exceptions or allow us in any way to get the details of the error condition?

---
In reply to: [205907511](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r205907511) [](ancestors = 205907511,203584606)

jennyf19 replied:
@SomkaPe Thanks for the feedback. I was concerned about that. will make the changes you suggested. 🐱

jennyf19 replied:
done.

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/BrokerHelper.cs:
```diff
@@ -210,6 +210,11 @@ public void AcquireToken(IDictionary<string, string> brokerPayload)
+                else
+                {
+                    throw new AdalException(AdalErrorAndroidEx.IncorrectBrokerRedirectUri, "The redirectUri is incorrect. The redirectUri should " +
+                        "be in this format: msauth://packagename/Base64UrlencodedSignature -> Check the ADAL logs for more details. ");
```
>Check the ADAL logs for more details. [](start = 94, length = 37)

what will actually be found in the logs?

jennyf19 replied:
I made a mistake...I added in logging here, which I didn't push up, but the [computed redirectUri is determined here](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/dev/adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/BrokerProxy.cs#L445). We could add logging in here, or i'll just change the above AdalException message. I thought we could actually log what the redirectUri should be, or we can just give the format. What do you think?

jennyf19 replied:
i fixed this.

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/BrokerHelper.cs:
```diff
@@ -210,6 +210,11 @@ public void AcquireToken(IDictionary<string, string> brokerPayload)
+                else
+                {
+                    throw new AdalException(AdalErrorAndroidEx.IncorrectBrokerRedirectUri, "The redirectUri is incorrect. The redirectUri should " +
+                        "be in this format: msauth://packagename/Base64UrlencodedSignature -> Check the ADAL logs for more details. ");
```
>  [](start = 131, length = 1)

empty...

jennyf19 replied:
not sure what you mean

jennyf19 replied:
still not sure what you mean here.

henrik-me commented:
approving PR expecting pointers raised to be addressed

henrik-me reviewed (approved):
:shipit:

henrik-me commented:
is something similar needed for the MSAL test app?

bgavrilMS reviewed (commented):
LGTM when dealt with Henrik's comments

jennyf19 commented:
we don't have broker support for MSAL (yet), but when we do, we should do something similar.

---
In reply to: [406158718](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#issuecomment-406158718) [](ancestors = 406158718)

jennyf19 committed:
add button for broker flow in adal test app

bgavrilMS commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -202,7 +208,54 @@ void LogCallback(LogLevel level, string message, bool containsPii)
             }
+        }
 
+        private async void AcquireTokenWithBrokerButton_Clicked(object sender, EventArgs e)
```
I believe Travis enabled some static analysis related to async. One of the rules is that all Async methods should be suffixed with Async.

henrik-me replied:
Perspective:
The rule is not meaningful for top level event handlers and can be ignored imo. This is the entry point for the async call. However either ignore and do not apply any async stuff to it or leave it with everything that the async rules says.

bgavrilMS replied:
Yes, agreed, especially as VS generates those even handlers from the designer. Irrespective of the solution chosen we should try not to introduce new warnings into the build. 

The number of warnings has dropped a lot in the past few weeks and once it reaches 0 we can enable "treat warnings as errors" and start looking at more static analysis.

jennyf19 replied:
I agree. I will leave this as-is. Thanks.

bgavrilMS commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -202,7 +208,54 @@ void LogCallback(LogLevel level, string message, bool containsPii)
+                    break;
+                case Device.Android:
+                    RedirectURI = AndroidBrokerRedirectURI;
+                    break;
```
Nit: I like to add a default: throw new NotImplementedException just in case ... fail fast if another runtime platform is used.

jennyf19 replied:
agree...sounds good. thanks

jennyf19 replied:
done.

bgavrilMS reviewed (approved):
LGTM, minor comments inline

jennyf19 committed:
Address PR Comments
- In ADAL dev app, include flows for broker interactive and silent. Remove clientIds and use variables throughout.
- For Android app, add broker parameter for `PlatformParameters`
- Fix broker error message
- pass in null if no idToken present for objectId
- remove identityProvider parameter as it is not used in this method

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/embeddedTestApp

jennyf19 commented:
@henrik-me @bgavrilMS 
While testing, silent flow w/Broker was not working, so I made additional changes. I have only tested w/Android w/Company Portal on a device. Both interactive and silent flows work.

TODO: iOS broker flows

jennyf19 commented on adal/devApps/XFormsApp.Droid/SecondPageRenderer.cs:
```diff
@@ -59,8 +49,14 @@ protected override void OnElementChanged(ElementChangedEventArgs<Page> e)
+            // Uncomment below for broker flow
+            //var parameters = new PlatformParameters(activity, true);
+            //page.Parameters = parameters;
+            #endregion
```
Not sure how to get around having to uncomment this to test broker flow. This is platform specific, so iOS has it's own implementation.

henrik-me replied:
Suggestion: add a button to the ui. for each platform do one or the other depending on the value of the current ui.

---
In reply to: [204958541](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r204958541) [](ancestors = 204958541)

jennyf19 replied:
went w/an idea from @SomkaPe

jennyf19 commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -37,10 +37,11 @@ public class SecondPage : ContentPage
+        public const string ClientIdBroker = "<ClientIdBroker>";
+        public const string AndroidBrokerRedirectURI = "msauth://com.microsoft.xformsdroid.adal/mJaAVvdXtcXy369xPWv2C7mV674=";
+        public const string IOSBrokerRedirectURI = "adaliosapp://com.yourcompany.xformsapp";
+        static string RedirectURI = "urn:ietf:wg:oauth:2.0:oob";
```
Using variables so the code is cleaner and easier to change if we need to. For using broker, the broker specific redirect uri needs to be registered in the app. I had to make my own clientId for testing this, but once we have one that we can use, we'll just register this redirectUri there.

jennyf19 commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs:
```diff
@@ -525,12 +525,18 @@ internal AdalResultWrapper LoadFromCacheCommon(CacheQueryData cacheQueryData, Re
+                if (idToken != null)
+                {
+                    objectId = idToken.ObjectId;
+                }
```
I was getting a null reference exception because the `IdToken` was null, which meant no `objectId` to pass through to `CacheFallBackOperations.WriteMsalRefreshToken(...)`. So i made a few changes to checkthis ahead of time and then just pass in null for `objectId`. Maybe @SomkaPe can confirm if it's okay to have a null `objectId` in the `MsalAccountCacheItem`.

SomkaPe replied:
Was it client credential grant case?

jennyf19 replied:
no...IDToken is not present in the current broker model

henrik-me replied:
is a dummy id needed in this case? @SomkaPe what will this do for the cache?

jennyf19 replied:
@henrik-me we got it figured out. it is okay to pass null here.

jennyf19 replied:
@SomkaPe - took your suggestion of doing this `idtoken?.ObjectId` instead of all the other logic. thanks.
And I re-tested interactive and silent flows w/broker w/and w/out a correct redirectUri and things work as expected.
cc: @henrik-me

jennyf19 commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs:
```diff
@@ -582,20 +588,20 @@ internal AdalResultWrapper LoadFromCacheCommon(CacheQueryData cacheQueryData, Re
-                                "A Multi Resource Refresh Token for a different resource was found which can be used";
-                            requestContext.Logger.Info(msg);
-                            requestContext.Logger.InfoPii(msg);
+                            // There are no resource specific tokens.  Choose any of the MRRT tokens if there are any.
```
I think `control + kd` changed the formatting on this.

jennyf19 commented on core/src/Cache/CacheFallbackOperations.cs:
```diff
@@ -38,7 +38,7 @@ internal class CacheFallbackOperations
         public static void WriteMsalRefreshToken(ITokenCacheAccessor tokenCacheAccessor,
             AdalResultWrapper resultWrapper, string authority, string clientId, string displayableId,
-            string identityProvider, string givenName, string familyName, string objectId)
+             string givenName, string familyName, string objectId)
```
`string identityProvider` wasn't being used anywhere in this method, so I just removed it. Doesn't make sense to pass in a parameter that is not used and `WriteMsalRefreshToken()` is only called here. We also don't have unit tests for this, but I need to learn more about the new cache format to write them.

jennyf19 committed:
fix format and add in clientId for easier testing until we use the scripts

henrik-me commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -29,23 +29,30 @@
     {
         private readonly StringBuilder _logs = new StringBuilder();
 
+        public const string ClientId = "d3590ed6-52b3-4102-aeff-aad2292ab01c";
```
>d3590ed6-52b3-4102-aeff-aad2292ab01c [](start = 40, length = 36)

did you intent to keep this id here?

jennyf19 replied:
yes.

henrik-me replied:
:shipit:

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/embeddedTestApp

jennyf19 committed:
Merge remote-tracking branch 'origin/dev' into jennyf/embeddedTestApp

jennyf19 committed:
addressing PR comments and using @pesomaka's suggestions for making code more readable and throwing exception sooner.

jennyf19 committed:
Merge branch 'dev' into jennyf/embeddedTestApp

jennyf19 committed:
Update for broker parameters for xamarin test app

henrik-me commented on adal/devApps/XFormsApp/SecondPage.cs:
```diff
@@ -195,21 +251,62 @@ void LogCallback(LogLevel level, string message, bool containsPii)
-                        new Uri("adaliosapp://com.yourcompany.xformsapp"),
-                        Parameters, new UserIdentifier("<USER>", UserIdentifierType.OptionalDisplayableId), null, claims).ConfigureAwait(false);
+                AuthenticationResult result = await ctx.AcquireTokenAsync("https://graph.windows.net", ClientId,
+                        new Uri(IOSBrokerRedirectURI),
```
is this intentional or did you mean to use the Device redirect uri?

jennyf19 replied:
that's what was there before, so i just replaced it w/what it had before.

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/BrokerProxy.cs:
```diff
@@ -364,17 +364,13 @@ public Intent GetIntentForBrokerActivity(AuthenticationRequest request, Activity
                 var ex = new AdalException("Authenticator cancels the request", e);
                 RequestContext.Logger.Error(ex);
                 RequestContext.Logger.ErrorPii(ex);
+                throw ex;
```
>throw ex; [](start = 16, length = 9)

was this a previous miss or is this an intentional behavior change? Is this handled anywhere in our code or do we pass this exception to the user?

jennyf19 replied:
@SomkaPe and I looked this over. It is an intentional behavior change. The exception is passed to the user.

henrik-me replied:
shouldn't this be in the above now removed catch intead. and we should leave the generic exception alone?

---
In reply to: [205926366](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r205926366) [](ancestors = 205926366)

jennyf19 replied:
The `OperationCancelledException` was just suppressed and written to the logs. I think we should catch the exceptions and throw an AdalException.

henrik-me replied:
yes, but you are now catching all exceptions and not just the specific one regarding operationcanceledexception... imo this code should be moved up to what was there before. and a general throw should be put in place for the general exception. (or it should not be handled at all)

---
In reply to: [205926683](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r205926683) [](ancestors = 205926683)

jennyf19 replied:
addressed all PR comments and tested

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/BrokerProxy.cs:
```diff
@@ -364,17 +364,13 @@ public Intent GetIntentForBrokerActivity(AuthenticationRequest request, Activity
             catch (Exception e)
             {
                 // Authenticator gets problem from webrequest or file read/write
                 var ex = new AdalException("Authenticator cancels the request", e);
```
>"Authenticator cancels the request" [](start = 43, length = 35)

imo, the exeptions are missing an id, or somehow it should be easy for the developer to do the right thing in case an exception occurs. Is this message available in a const anywhere?
should it be cancelled instead?

jennyf19 replied:
it's coming from the broker intent, so it would make sense to make it a const message, but leave as is. our cancelled one is for user cancelling auth, which is not the case here.

jennyf19 replied:
done.

henrik-me replied:
Cancels ... is that the right wording?

---
In reply to: [205926754](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1140#discussion_r205926754) [](ancestors = 205926754)

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/BrokerProxy.cs:
```diff
@@ -364,17 +364,13 @@ public Intent GetIntentForBrokerActivity(AuthenticationRequest request, Activity
-            }
             catch (Exception e)
             {
                 // Authenticator gets problem from webrequest or file read/write
```
>// Authenticator gets problem from webrequest or file read/write [](start = 16, length = 64)

is this comment relevant? #Resolved

jennyf19 replied:
maybe...maybe not...doesn't seem relevant now. ⛄️

jennyf19 replied:
removed.

jennyf19 committed:
address PR comments

jennyf19 committed:
fixing exceptions
